File: backend-legacy-python/app/vpn/protocols/vless_reality.py
"""VLESS Reality — TCP (Vision flow) + XHTTP + gRPC inbounds.

Note: XHTTP (port 2097) is intentionally NOT included in client_links() because
sing-box does not support XHTTP transport. XHTTP inbounds are still generated
for xray-native clients and node configs.
"""

import hashlib
from urllib.parse import quote

from app.config import get_settings

from .base import (
    ClientLink,
    Protocol,
    ServerConfig,
    SingboxOutbound,
    UserCredentials,
    XrayInbound,
)

SNIFFING = {"enabled": True, "destOverride": ["http", "tls"]}
FINGERPRINT = "chrome"


class VlessReality(Protocol):
    def __init__(self) -> None:
        s = get_settings()
        self._private_key = s.reality_private_key
        self._public_key = s.reality_public_key
        self._snis = s.reality_snis or ["ads.x5.ru"]
        self._dest = f"{self._snis[0]}:443"
        self._tcp_port = s.vless_tcp_port
        self._grpc_port = s.vless_grpc_port
        self._relays = s.relay_servers
        self._xray_version = s.xray_version

    @property
    def name(self) -> str:
        return "vless_reality"

    @property
    def display_name(self) -> str:
        return "VLESS Reality"

    # ── Helpers ──

    def _reality_settings(self, short_ids: list[str]) -> dict:
        return {
            "show": False,
            "dest": self._dest,
            "xver": 0,
            "serverNames": self._snis,
            "privateKey": self._private_key,
            "shortIds": sorted(set(short_ids) | {""}),
            "echForceQuery": "full",  # xray v26.3.27+: force ECH query for all connections
        }

    def _make_inbound(
        self, tag: str, port: int, network: str, clients: list, short_ids: list[str]
    ) -> XrayInbound:
        stream: dict = {
            "network": network,
            "security": "reality",
            "realitySettings": self._reality_settings(short_ids),
        }
        if network == "tcp":
            stream["sockopt"] = {"tcpFastOpen": True}
        elif network == "grpc":
            stream["grpcSettings"] = {"serviceName": ""}
        elif network == "xhttp":
            stream["xhttpSettings"] = {
                "mode": "auto",
                "browserMasquerading": "chrome",  # xray v26.3.27+: disguise as Chrome browser
            }

        settings: dict = {"clients": clients, "decryption": "none"}
        if network == "tcp":
            settings["fallbacks"] = []

        return XrayInbound(
            tag=tag, port=port, protocol="vless",
            settings=settings, stream_settings=stream, sniffing=SNIFFING,
        )

    def _build_clients(
        self, users: list[UserCredentials], suffix: str, flow: str = ""
    ) -> list[dict]:
        clients = []
        for u in users:
            c: dict = {"id": u.uuid, "email": f"{u.username}@{suffix}"}
            if flow:
                c["flow"] = flow
            clients.append(c)
        return clients

    def _get_user_snis(self, username: str, count: int = 5) -> list[str]:
        # Use SHA-256 for stable hash across Python restarts (hash() is randomized)
        offset = hashlib.sha256(username.encode()).digest()[0] % len(self._snis)
        rotated = self._snis[offset:] + self._snis[:offset]
        return rotated[:count]

    # ── Protocol interface ──

    def xray_inbounds(
        self, users: list[UserCredentials], short_ids: list[str]
    ) -> list[XrayInbound]:
        tcp_clients = self._build_clients(users, "xray", "xtls-rprx-vision")
        xhttp_clients = self._build_clients(users, "xhttp")
        grpc_clients = self._build_clients(users, "grpc")
        return [
            self._make_inbound("VLESS TCP REALITY", self._tcp_port, "tcp", tcp_clients, short_ids),
            self._make_inbound("VLESS XHTTP REALITY", 2097, "xhttp", xhttp_clients, short_ids),
            self._make_inbound("VLESS gRPC REALITY", self._grpc_port, "grpc", grpc_clients, short_ids),
        ]

    def client_links(
        self, user: UserCredentials, servers: list[ServerConfig]
    ) -> list[ClientLink]:
        links: list[ClientLink] = []
        user_snis = self._get_user_snis(user.username)
        default_sni = user_snis[0]

        # Relay links (LTE whitelist bypass)
        for relay in self._relays:
            relay_host = relay.get("relay_domain", relay["relay_ip"])
            for target in relay["targets"]:
                base = f"{target['flag']} {target['name']} LTE"
                links.append(self._tcp_link(user, relay_host, target["tcp_port"], default_sni, f"{base} TCP", is_relay=True))
                links.append(self._grpc_link(user, relay_host, target["grpc_port"], default_sni, f"{base} gRPC", is_relay=True))

        # Direct server links
        for srv in servers:
            host = srv.domain or srv.host
            base = f"{srv.flag} {srv.name}"
            for sni in user_snis:
                links.append(self._tcp_link(user, host, self._tcp_port, sni, f"{base} [{sni.split('.')[0]}]", server_key=srv.key))
            links.append(self._grpc_link(user, host, self._grpc_port, default_sni, f"{base} gRPC", server_key=srv.key))

        return links

    def singbox_outbound(
        self, tag: str, server: ServerConfig, user: UserCredentials, **opts
    ) -> dict | None:
        transport = opts.get("transport", "tcp")
        sni = opts.get("sni", self._snis[0])
        out: dict = {
            "type": "vless",
            "tag": tag,
            "server": server.domain or server.host,
            "server_port": self._tcp_port if transport == "tcp" else self._grpc_port,
            "uuid": user.uuid,
            "tls": {
                "enabled": True,
                "server_name": sni,
                "utls": {"enabled": True, "fingerprint": FINGERPRINT},
                "reality": {"enabled": True, "public_key": self._public_key, "short_id": user.short_id},
            },
        }
        if transport == "tcp":
            out["flow"] = "xtls-rprx-vision"
        elif transport == "grpc":
            out["transport"] = {"type": "grpc", "service_name": ""}

        # SMUX/Multiplex — reduces handshake overhead, adds padding
        out["multiplex"] = {
            "enabled": True,
            "protocol": "h2mux",
            "max_connections": 4,
            "padding": True,
        }
        return out

    # ── Link builders ──

    def _tcp_link(
        self, user: UserCredentials, host: str, port: int, sni: str, remark: str,
        server_key: str = "relay", is_relay: bool = False,
    ) -> ClientLink:
        uri = (
            f"vless://{user.uuid}@{host}:{port}"
            f"?type=tcp&security=reality&sni={sni}&fp={FINGERPRINT}"
            f"&pbk={self._public_key}&sid={user.short_id}&flow=xtls-rprx-vision"
            f"#{quote(remark)}"
        )
        return ClientLink(uri=uri, protocol="vless", transport="tcp", server_key=server_key, remark=remark, is_relay=is_relay)

    def _grpc_link(
        self, user: UserCredentials, host: str, port: int, sni: str, remark: str,
        server_key: str = "relay", is_relay: bool = False,
    ) -> ClientLink:
        uri = (
            f"vless://{user.uuid}@{host}:{port}"
            f"?type=grpc&security=reality&sni={sni}&fp={FINGERPRINT}"
            f"&pbk={self._public_key}&sid={user.short_id}"
            f"&serviceName=&authority=&encryption=none"
            f"#{quote(remark)}"
        )
        return ClientLink(uri=uri, protocol="vless", transport="grpc", server_key=server_key, remark=remark, is_relay=is_relay)

File: backend/crates/chameleon-vpn/src/protocols/vless_reality.rs
//! VLESS Reality — TCP (Vision flow) + XHTTP + gRPC inbounds.

use sha2::{Sha256, Digest};
use serde_json::json;

use chameleon_config::Settings;
use super::types::*;

const FINGERPRINT: &str = "chrome";

pub struct VlessReality {
    private_key: String,
    public_key: String,
    snis: Vec<String>,
    dest: String,
    tcp_port: u16,
    grpc_port: u16,
    relays: String, // raw JSON from config
}

impl VlessReality {
    pub fn new(s: &Settings) -> Self {
        let snis = if s.reality_snis.is_empty() { vec!["ads.x5.ru".into()] } else { s.reality_snis.clone() };
        let dest = format!("{}:443", snis[0]);
        Self {
            private_key: s.reality_private_key.clone(),
            public_key: s.reality_public_key.clone(),
            snis,
            dest,
            tcp_port: s.vless_tcp_port,
            grpc_port: s.vless_grpc_port,
            relays: s.relay_servers_raw.clone(),
        }
    }

    fn reality_settings(&self, short_ids: &[String]) -> serde_json::Value {
        let mut ids: Vec<String> = short_ids.to_vec();
        if !ids.contains(&String::new()) { ids.push(String::new()); }
        ids.sort();
        ids.dedup();
        json!({
            "show": false,
            "dest": self.dest,
            "xver": 0,
            "serverNames": self.snis,
            "privateKey": self.private_key,
            "shortIds": ids,
            "echForceQuery": "full",
        })
    }

    fn make_inbound(&self, tag: &str, port: u16, network: &str, clients: Vec<serde_json::Value>, short_ids: &[String]) -> XrayInbound {
        let mut stream = json!({
            "network": network,
            "security": "reality",
            "realitySettings": self.reality_settings(short_ids),
        });
        match network {
            "tcp" => { stream["sockopt"] = json!({"tcpFastOpen": true}); }
            "grpc" => { stream["grpcSettings"] = json!({"serviceName": ""}); }
            "xhttp" => { stream["xhttpSettings"] = json!({"mode": "auto", "browserMasquerading": "chrome"}); }
            _ => {}
        }

        let mut settings = json!({"clients": clients, "decryption": "none"});
        if network == "tcp" {
            settings["fallbacks"] = json!([]);
        }

        XrayInbound {
            tag: tag.to_string(),
            port,
            protocol: "vless".to_string(),
            settings,
            stream_settings: stream,
            sniffing: json!({"enabled": true, "destOverride": ["http", "tls"]}),
            listen: "0.0.0.0".to_string(),
        }
    }

    fn build_clients(&self, users: &[UserCredentials], suffix: &str, flow: &str) -> Vec<serde_json::Value> {
        users.iter().map(|u| {
            let mut c = json!({"id": u.uuid, "email": format!("{}@{}", u.username, suffix)});
            if !flow.is_empty() { c["flow"] = json!(flow); }
            c
        }).collect()
    }

    fn get_user_snis(&self, username: &str, count: usize) -> Vec<String> {
        if self.snis.is_empty() { return vec![]; }
        let hash = Sha256::digest(username.as_bytes());
        let offset = (hash[0] as usize) % self.snis.len();
        let mut rotated: Vec<String> = self.snis[offset..].iter().chain(self.snis[..offset].iter()).cloned().collect();
        rotated.truncate(count);
        rotated
    }

    fn tcp_link(&self, user: &UserCredentials, host: &str, port: u16, sni: &str, remark: &str, server_key: &str, is_relay: bool) -> ClientLink {
        let uri = format!(
            "vless://{}@{}:{}?type=tcp&security=reality&sni={}&fp={}&pbk={}&sid={}&flow=xtls-rprx-vision#{}",
            user.uuid, host, port, sni, FINGERPRINT, self.public_key, user.short_id, urlencoding::encode(remark)
        );
        ClientLink { uri, protocol: "vless".into(), transport: "tcp".into(), server_key: server_key.into(), remark: remark.into(), is_relay }
    }

    fn grpc_link(&self, user: &UserCredentials, host: &str, port: u16, sni: &str, remark: &str, server_key: &str, is_relay: bool) -> ClientLink {
        let uri = format!(
            "vless://{}@{}:{}?type=grpc&security=reality&sni={}&fp={}&pbk={}&sid={}&serviceName=&authority=&encryption=none#{}",
            user.uuid, host, port, sni, FINGERPRINT, self.public_key, user.short_id, urlencoding::encode(remark)
        );
        ClientLink { uri, protocol: "vless".into(), transport: "grpc".into(), server_key: server_key.into(), remark: remark.into(), is_relay }
    }
}

impl Protocol for VlessReality {
    fn name(&self) -> &str { "vless_reality" }
    fn display_name(&self) -> &str { "VLESS Reality" }
    fn port(&self) -> u16 { self.tcp_port }

    fn xray_inbounds(&self, users: &[UserCredentials], short_ids: &[String]) -> Vec<XrayInbound> {
        let tcp = self.build_clients(users, "xray", "xtls-rprx-vision");
        let xhttp = self.build_clients(users, "xhttp", "");
        let grpc = self.build_clients(users, "grpc", "");
        vec![
            self.make_inbound("VLESS TCP REALITY", self.tcp_port, "tcp", tcp, short_ids),
            self.make_inbound("VLESS XHTTP REALITY", 2097, "xhttp", xhttp, short_ids),
            self.make_inbound("VLESS gRPC REALITY", self.grpc_port, "grpc", grpc, short_ids),
        ]
    }

    fn client_links(&self, user: &UserCredentials, servers: &[ServerConfig]) -> Vec<ClientLink> {
        let mut links = vec![];
        let user_snis = self.get_user_snis(&user.username, 5);
        let default_sni = user_snis.first().map(|s| s.as_str()).unwrap_or("ads.x5.ru");

        for srv in servers {
            let host = srv.effective_host();
            let base = format!("{} {}", srv.flag, srv.name);
            for sni in &user_snis {
                let sni_label = sni.split('.').next().unwrap_or(sni);
                links.push(self.tcp_link(user, host, self.tcp_port, sni, &format!("{base} [{sni_label}]"), &srv.key, false));
            }
            links.push(self.grpc_link(user, host, self.grpc_port, default_sni, &format!("{base} gRPC"), &srv.key, false));
        }
        links
    }

    fn singbox_outbound(&self, tag: &str, server: &ServerConfig, user: &UserCredentials, opts: &OutboundOpts) -> Option<serde_json::Value> {
        let transport = opts.transport.as_deref().unwrap_or("tcp");
        let sni = opts.sni.as_deref().unwrap_or_else(|| self.snis.first().map(|s| s.as_str()).unwrap_or("ads.x5.ru"));
        let host = server.effective_host();
        let port = if transport == "tcp" { self.tcp_port } else { self.grpc_port };

        let mut out = json!({
            "type": "vless",
            "tag": tag,
            "server": host,
            "server_port": port,
            "uuid": user.uuid,
            "tls": {
                "enabled": true,
                "server_name": sni,
                "utls": {"enabled": true, "fingerprint": FINGERPRINT},
                "reality": {"enabled": true, "public_key": self.public_key, "short_id": user.short_id},
            },
        });
        if transport == "tcp" {
            // flow and multiplex are mutually exclusive in sing-box — Vision operates at TLS layer
            out["flow"] = json!("xtls-rprx-vision");
        } else if transport == "grpc" {
            out["transport"] = json!({"type": "grpc", "service_name": ""});
            // gRPC has built-in multiplexing, adding h2mux on top causes issues
        }
        Some(out)
    }
}
